---
title: "tutorial"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{tutorial}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

### Overview

This tutorial demonstrates how to use Longcell to analyze single cell isoform count data. Due to the frequent truncations in Nanopore long reads sequencing and to allow for identification of new ioforms, Longcell stores the isoform information as sequence of exons. So instead of building a cell-by-isoform matrix, the information of gene expression and isoform expression are stored seperately. 

The general steps for the analysis are:

1. regular scRNA-seq gene expression analysis.

2. identify highly variable exons

3. differetial alternative splicing analysis between cell groups

We add interface to Seurat, thus you can do both general scRNA-seq gene expression analysis and alternative splicing analysis within a Seurat object.

First, we load Longcell and the other packages necessary for this vignette.

```{r setup}
library(Longcell)
library(Seurat)
library(dplyr)
```

### annotation

```{r}
gene_bed = readRDS("../tests/annotation/gene_bed.rds")
gtf = readRDS("../tests/annotation/exon_gtf.rds")
```

### data

Here we are using the sample of colorecatal metastasis to liver published in our paper as an example. The single cell isoform quantification for this sample has been generated by Longcell-pre.

We can directly generate a Seurat object from the output of Longcell-pre.The Splice object is embedded in the `misc` slot of the Seurat object, which stores the single cell isoform count. You can get access to it via `getSplice(object)`

```{r}
crc = createSpliceSeurat(path = "../tests/data/",project = "colorectal",min.cells = 3,min.features = 200)
```

```{r}
getSplice(crc)
```

We could check the isoform count for each gene in each cell via `getIsoform(object,gene)`, for example:

```{r}
crc_MYL6_iso = getIsoform(crc,"MYL6")
head(crc_MYL6_iso)
```

We can first do general single cell gene count analysis via regular Seurat process to get the cell cluster information

```{r}
crc <- PercentageFeatureSet(crc, pattern = "^MT-", col.name = "percent.mt")
```

```{r}
crc <- SCTransform(crc, method = "glmGamPoi",vars.to.regress = "percent.mt", verbose = FALSE)
```

```{r}
crc <-  crc %>%
    RunPCA(npcs = 50, verbose = FALSE,approx = F) %>%
    RunUMAP(reduction = "pca", dims = 1:30,verbose = FALSE) %>% 
    FindNeighbors(reduction = "pca", dims = 1:30,verbose = FALSE) %>% 
    FindClusters(resolution = 1.5,verbose = FALSE) %>% 
    identity()
```

Here we directly use the cell type annotation we get from marker genes to annotate cell clusters.

```{r}
cell_type = read.table("../tests/data/cell_type.txt",header = TRUE)
rownames(cell_type) = cell_type$cell
crc$type = cell_type[colnames(crc),"type"]
```

```{r}
DimPlot(crc, reduction = "umap", group.by = "type",pt.size = 1)
```

### Single cell heterogeneity of alterative splicing

Alternative splicing heterogeneity can only be confident estimated for genes with high expression in multiple cells, a common threshold is at least 10 transcripts in over 30 cells, thus we could first filter out low expressed genes via their total expression.

```{r}
heg = HighExprsGene(crc,thresh = 500,slot = "RNA")
```

```{r}
head(heg)
length(heg)
```

Here 355 genes can be preserved for downstream $\phi$ calculation. The gene bed and exon gtf annotations are optional but recommened. Gene bed annotations can be used to filter out signals from too short exons, which are volunarable to sequencing and mapping errors. Gtf annotation can help merge exons which are always coexisting or mutually exclusive, removing replicative signals. 

```{r}
crc_phi = genes_exons_phi(crc,genes = heg,gene_bed = gene_bed,gtf = gtf,cores = 8)
```

We can then show all valid $\phi$ estimation with small confidence interval in a $\phi$ vs. $\bar{\psi}$ scatter plot

```{r}
phiPlot(crc_phi,annot_col = "gene",midpoint = 0.1)
```

We could order the $\phi$ for each exon in a decreasing order, and the top list should be highly variable exons.

```{r}
head(crc_phi[order(crc_phi$phi,decreasing = TRUE),])
```

Here we use the highest one MYL6 exon35 as an example, we could first show its single cell $\psi$ distribution:

```{r}
psiHist(crc,gene = "MYL6",exon = "35")
```

```{r}
psiCellPlot(crc,gene = "MYL6",exons = "35",cell_embedding = "umap")
```

We could see very different alternative splicing patterns for this exon in epithlials and other immune cells. Epithlials highly spliced in the exon 35 in MYL6, while T cells and macrophages tend to splice this exon out.

### Generalized Likelihood Ratio Test

Since we have found the splicing heterogeneity between epithelials and immune cells, here we could do a differential alternative splicig analysis between the two cell groups to see if they have other more splicing difference.

We first specify the non-epithelial cells as immune cells in the cell type annotation

```{r}
crc$group = ifelse(crc$type == "epithelial","epithelial","immune")
```

Then we could do generalized likelihood ratio test for highly expressed genes for immune cells and epithelials to see if they have alternative splicing difference.

```{r}
crc_GLRT_sig = genes_groups_GLRT(crc,genes = heg,
                              group1s = "epithelial",group2s = "immune",group_col = "group",
                              gene_bed = gene_bed,gtf = gtf,q_thresh = 0.05,cores = 8)
```

Compared to epthelials, most exons show decreasing splice-in in immune cells as most of them show negative mean change.

```{r}
GLRT_sig_plot(crc_GLRT_sig)
```

```{r}
head(crc_GLRT_sig[order(crc_GLRT_sig$q),])
```

We could first check the $\psi$ distribution for MYL6 exon 35 in two cell groups as it's the most significant signal.

```{r}
psiHist(crc,gene = "MYL6",exon = "35",group_col = "group",group = "epithelial",
           alpha = 4.585,beta = 3.364)
psiHist(crc,gene = "MYL6",exon = "35",group_col = "group",group = "immune",
            alpha = 1.296922,beta = 6.654382)
```

We could also check other genes, for example, RPS24:

```{r}
psiHist(crc,gene = "RPS24",exon = "31",group_col = "group",group = "epithelial",
           alpha = 6.683290,beta = 3.766842)
psiHist(crc,gene = "RPS24",exon = "31",group_col = "group",group = "immune",
            alpha = 5.211762,beta = 7.757096)
```

