---
title: "Tutorial for Longcell"
output: rmarkdown::html_vignette
vignette: |
  index: true
---

### Overview

This tutorial demonstrates how to use Longcell to analyze single cell isoform count data. Due to the frequent truncations in Nanopore long reads sequencing and to allow for identification of new ioforms, Longcell stores the isoform information as sequence of splicing sites. So instead of building a cell-by-isoform matrix, Longcell reveal alternative splicing events on the splicing sites level. But the cell-by-isoform matrix is also provided for regular single cell RNA seq analysis,like cell type clustering. 

The general steps for the analysis are:

1. regular scRNA-seq gene/isoform expression analysis.

2. identify highly variable splicing sites.

3. differetial alternative splicing analysis between cell groups

First, we load Longcell and the other packages necessary for this vignette.

```{r}
options(future.globals.maxSize= 1024*1024^2)
```

```{r}
library(Longcell)
library(Seurat)
library(dplyr)
library(future)
library(future.apply)
library(ggplot2)
```

### annotation

The gene strand information is necessary to build meta splicing sites in the downstream analysis, here we can use the `gene_bed.rds` output from the LongcellPre package, but you can also use your annotation file to indicate the gene strand information. It should be a table with at least two columns, one for the gene name, and the other one for the gene strand.

```{r}
gene_bed = readRDS("../inst/extdata//annotation//gene_bed.rds")
```

### data overview

Here we are using the sample of colorecatal metastasis to liver published in our paper as an example. The single cell isoform quantification for this sample is generated by LongcellPre.

We can generate a Seurat object based on the cell by isoform matrix, just in the same way as the cell by gene matrix.

```{r}
iso_mat = Read10X("../inst/extdata/data/isoform/",gene.column = 1)
```

```{r}
crc = CreateSeuratObject(counts = iso_mat,project = "isoform",min.cells = 3,min.features = 200)
```

```{r}
crc
```

We can first do general single cell analysis via regular Seurat process to get the cell cluster information

```{r}
suppressWarnings({
    crc <-  crc %>%
    SCTransform(method = "glmGamPoi", verbose = FALSE) %>%
    RunPCA(npcs = 30, verbose = FALSE) %>%
    RunUMAP(reduction = "pca", dims = 1:20,n.neighbors = 30,verbose = FALSE)
})
```

Here we directly use the cell type annotation we get from marker genes to annotate cell clusters.

```{r}
cell_type = read.table("../inst/extdata/data/cell_type.txt",header = TRUE)
rownames(cell_type) = cell_type$cell
crc$type = cell_type[colnames(crc),"type"]
```

```{r}
DimPlot(crc, reduction = "umap", group.by = "type",label = TRUE,pt.size = 2)+NoLegend()
```

### Single cell heterogeneity of alterative splicing

After we take a glance on the data, we can start to explore the alternative splicing within the cell population. Longcell detect the alternative splicing event at the splice site level.We first read in the data for the downstream analysis.

```{r}
iso = readRDS("../inst/extdata/data/sc_iso_count.rds")
```

Each read here is stored as $s_1,e_1|s_2,e_2|...$, in which $s_i$ means the start position of the exon $i$ while $e_i$ means the end position.

We first build a Splice object based on this data.

```{r}
crc_splice = creatSplice_from_df(iso)
```

The isoform sequences and expression can be revisted from the Splice object by `getIsoform()`

```{r}
temp = getIsoform(crc_splice,c("MYL6","RPS24"))
```

```{r}
head(temp)
```

As isoform count is even sparser than the gene count, we can only get confident alternative splicing signals from high expression genes, thus we can filter out lowly expressed genes to save the computing time.

```{r}
heg = HighExprsGene(crc_splice,thresh = 500)
head(heg)
length(heg)
```

Here 752 genes would be preserved for downstream $\phi$ calculation. And before that, we can first merge splicing sites which are always coexisting or mutually exclusive as meta splicing sites, as they contain the same information. Through this step we can further decrease the computation burden. The parameter `eps` can be adjusted to indicate how aggressive the merging process would be.Higher `eps` leads to more aggressive merge, but may be more possible to lead to information loss.

```{r}
plan(strategy = "multisession",workers = 8)
crc_splice = geneSiteTable(crc_splice,gene_bed,genes = heg,eps = 0.05,verbose = FALSE)
```

The merged splicing sites are saved in a `metaSite` object, for example:

```{r}
getMetaSites(crc_splice,"MYL6")
```

The `metaSite` object maintains two matrices, one records the gene count for each meta splice site and the other records the spliced-in count. The correspondence between the meta sites and original sites is also recorded in it.

After we generate the meta splicing sites, we can estimate the ratio of inter-cell heterogeneity for each meta site.

```{r}
plan(strategy = "multisession",workers = 8)
crc_phi = genesSitesPhi(crc_splice,genes = heg,method = "beta",verbose = FALSE)
```

Each row in the result table records the $\bar{\psi}$ (percent-spliced-in) and $\phi$ (inter-cell heterogeneity) estimation and the confidence interval for $\phi$. The count `column` means the number of cells which have valid gene expression to estimate the $\psi$ and $\phi$.

```{r}
head(crc_phi)
```

We can then show all valid $\phi$ estimation with small confidence interval in a $\phi$ vs. $\bar{\psi}$ scatter plot

```{r}
phiPlot(crc_phi,annot_col = "gene")
```

We could order the $\phi$ for each meta splicing site in a decreasing order, and the top list should be highly variable exons.

```{r}
head(crc_phi[order(crc_phi$phi,decreasing = TRUE),])
```

Here we use the highest one MYL6 metasite_4 as an example, we could first show its single cell $\psi$ distribution,the color of the bar indicates the average gene expression:

```{r}
gene = "MYL6"
site = "metasite_4"
```

```{r}
psiHist(crc_splice,gene = gene,site = site)
```

We could also show the $\psi$ distribution in the cell embedding umap:

```{r}
embedding = as.data.frame(crc@reductions$umap@cell.embeddings)
```

```{r}
psiCellPlot(crc_splice,gene = gene,site = site,cell_embedding = embedding)
```

We could see very different alternative splicing patterns for this exon in epithlials and other immune cells. Epithlials highly spliced in the metasite_4 in MYL6, while T cells and macrophages tend to splice this meta site out.

If we trace back to the original sites, we could find that the metasite_4 contains two sites:

```{r}
getMetaSites(crc_splice,"MYL6")@sites
```

We can compare the sashimi plot between the epithlials and the immune cells group to see which exons are differetially spliced between the two cell groups.

```{r}
epithlials = colnames(crc)[crc$type == "epithelial"]
immune = colnames(crc)[crc$type != "epithelial"]
```

```{r}
sashimi(crc_splice,gene = "MYL6",cells = epithlials)+xlim(c(56158300,56161500))
```

```{r}
sashimi(crc_splice,gene = "MYL6",cells = immune,color_id = 2,filter_ratio = 10)+xlim(c(56158300,56161500))
```

From the sashimi plot we can see that the exon $6$ has differetial alternative splicing between the epithlials and immune cells.

### Generalized Likelihood Ratio Test

Since we have found the splicing heterogeneity between epithelials and immune cells, here we could do a differential alternative splicig analysis between the two cell groups to see if they have other more splicing difference.

Still, we would do generalized likelihood ratio test for highly expressed genes.

```{r}
plan(strategy = "multisession",workers = 8)
group1 = list(epithlials)
group2 = list(immune)
names(group1) = "epithlials"
names(group2) = "immune"
crc_GLRT_sig = genes_groups_GLRT(crc_splice,genes = names(crc_splice@meta_sites),
                              group1s = list(epithlials),group2s = list(immune))
```

```{r}
plan(strategy = "multisession",workers = 8)
crc_GLRT_sig = suppressWarnings(genes_multigroups_GLRT(spliceOb = crc_splice,
                                      genes = names(crc_splice@meta_sites),
                                      meta = cell_type,group_col = "type"))
```

```{r}
head(crc_GLRT_sig[order(crc_GLRT_sig$q),])
```

```{r}
GLRT_sig_plot(crc_GLRT_sig,color_col = "group")
```

The metasite_4 in MYL6 is one of the most significant result in the differential alternative splicing analysis between the immune and epithelial cells, which follows our expectation. As from last section, we do observe obvious inter-cell heterogeneity for this metasite.

Besides MYL6, there also other metasites show siginificant difference. For example, the metasite_3 in RPS24, we can observe its $\psi$ distribtion in two cell populations as a sanity check.

```{r}
target_group = "epithelial"
target_gene = "RPS24"
target_site = "metasite_3"
target = crc_GLRT_sig %>% filter(group == target_group,gene == target_gene,site == target_site)
```

```{r}
target_cells = rownames(cell_type)[cell_type$type == target_group]
control_cells = rownames(cell_type)[cell_type$type != target_group]
```

```{r}
psiHist(crc_splice,gene = target_gene,site = target_site,cells = target_cells,
        alpha = target$alpha1,beta = target$beta1)
psiHist(crc_splice,gene = target_gene,site = target_site,cells = control_cells,
        alpha = target$alpha2,beta = target$beta2)
```

From above histogram we do see, compared to the immune cells, RPS24 in epthelial preserved more metasite_3.

```{r}
getMetaSites(crc_splice,target_gene)@sites
```

```{r}
sashimi(crc_splice,gene = target_gene,cells = target_cells)
```

```{r}
sashimi(crc_splice,gene = target_gene,cells = control_cells,color_id = 2)
```

