<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Longcell">
<title>tutorial • Longcell</title>
<script src="../deps/jquery-3.5.1/jquery-3.5.1.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="tutorial">
<meta property="og:description" content="Longcell">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">Longcell</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/Longcell.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/getStart.html">tutorial</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>tutorial</h1>
            
      
      
      <div class="d-none name"><code>getStart.Rmd</code></div>
    </div>

    
    
<div class="section level3">
<h3 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h3>
<p>This tutorial demonstrates how to use Longcell to analyze single cell isoform count data. Due to the frequent truncations in Nanopore long reads sequencing and to allow for identification of new ioforms, Longcell stores the isoform information as sequence of exons. So instead of building a cell-by-isoform matrix, the information of gene expression and isoform expression are stored seperately.</p>
<p>The general steps for the analysis are:</p>
<ol style="list-style-type: decimal">
<li><p>regular scRNA-seq gene expression analysis.</p></li>
<li><p>identify highly variable exons</p></li>
<li><p>differetial alternative splicing analysis between cell groups</p></li>
</ol>
<p>We add interface to Seurat, thus you can do both general scRNA-seq gene expression analysis and alternative splicing analysis within a Seurat object.</p>
<p>First, we load Longcell and the other packages necessary for this vignette.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">Longcell</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Attaching SeuratObject</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="annotation">annotation<a class="anchor" aria-label="anchor" href="#annotation"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene_bed</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"../tests/annotation/gene_bed.rds"</span><span class="op">)</span></span>
<span><span class="va">gtf</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"../tests/annotation/exon_gtf.rds"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="data">data<a class="anchor" aria-label="anchor" href="#data"></a>
</h3>
<p>Here we are using the sample of colorecatal metastasis to liver published in our paper as an example. The single cell isoform quantification for this sample has been generated by Longcell-pre.</p>
<p>We can directly generate a Seurat object from the output of Longcell-pre.The Splice object is embedded in the <code>misc</code> slot of the Seurat object, which stores the single cell isoform count. You can get access to it via <code>getSplice(object)</code></p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">crc</span> <span class="op">=</span> <span class="fu"><a href="../reference/createSpliceSeurat.html">createSpliceSeurat</a></span><span class="op">(</span>path <span class="op">=</span> <span class="st">"../tests/data/"</span>,project <span class="op">=</span> <span class="st">"colorectal"</span>,min.cells <span class="op">=</span> <span class="fl">3</span>,min.features <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/getSplice.html">getSplice</a></span><span class="op">(</span><span class="va">crc</span><span class="op">)</span></span>
<span><span class="co">#&gt; Splice object with  239  cells and  14733 genes</span></span></code></pre></div>
<p>We could check the isoform count for each gene in each cell via <code>getIsoform(object,gene)</code>, for example:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">crc_MYL6_iso</span> <span class="op">=</span> <span class="fu"><a href="../reference/getIsoform.html">getIsoform</a></span><span class="op">(</span><span class="va">crc</span>,<span class="st">"MYL6"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">crc_MYL6_iso</span><span class="op">)</span></span>
<span><span class="co">#&gt;                  cell gene    start</span></span>
<span><span class="co">#&gt; 2729 AAACGGGGTTATGTGC MYL6 56158371</span></span>
<span><span class="co">#&gt; 2730 AAACGGGGTTATGTGC MYL6 56158371</span></span>
<span><span class="co">#&gt; 2731 AAACGGGGTTATGTGC MYL6 56160017</span></span>
<span><span class="co">#&gt; 2732 AAACGGGGTTATGTGC MYL6 56160017</span></span>
<span><span class="co">#&gt; 2733 AAACGGGGTTATGTGC MYL6 56158359</span></span>
<span><span class="co">#&gt; 2734 AAACGGGGTTATGTGC MYL6 56158359</span></span>
<span><span class="co">#&gt;                                                              exons      end</span></span>
<span><span class="co">#&gt; 2729               11|12|13|14|17|22|23|24|27|28|29|31|32|39|40|41 56161573</span></span>
<span><span class="co">#&gt; 2730            11|12|13|14|17|22|23|24|27|28|29|31|32|39|40|41|42 56161577</span></span>
<span><span class="co">#&gt; 2731                                          28|29|31|32|39|40|41 56161572</span></span>
<span><span class="co">#&gt; 2732                                          28|29|31|32|39|40|41 56161575</span></span>
<span><span class="co">#&gt; 2733 6|7|8|9|10|11|12|13|14|17|22|23|24|27|28|29|31|32|35|39|40|41 56161574</span></span>
<span><span class="co">#&gt; 2734          6|7|8|9|10|11|12|13|14|17|22|23|24|27|28|29|31|32|39 56161473</span></span>
<span><span class="co">#&gt;      count     polyA</span></span>
<span><span class="co">#&gt; 2729     0 1.0000000</span></span>
<span><span class="co">#&gt; 2730     1 1.0000000</span></span>
<span><span class="co">#&gt; 2731     0 1.0000000</span></span>
<span><span class="co">#&gt; 2732     1 0.9736842</span></span>
<span><span class="co">#&gt; 2733     1 0.9722222</span></span>
<span><span class="co">#&gt; 2734     0 1.0000000</span></span></code></pre></div>
<p>We can first do general single cell gene count analysis via regular Seurat process to get the cell cluster information</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">crc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/PercentageFeatureSet.html" class="external-link">PercentageFeatureSet</a></span><span class="op">(</span><span class="va">crc</span>, pattern <span class="op">=</span> <span class="st">"^MT-"</span>, col.name <span class="op">=</span> <span class="st">"percent.mt"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">crc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SCTransform.html" class="external-link">SCTransform</a></span><span class="op">(</span><span class="va">crc</span>, method <span class="op">=</span> <span class="st">"glmGamPoi"</span>,vars.to.regress <span class="op">=</span> <span class="st">"percent.mt"</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">crc</span> <span class="op">&lt;-</span>  <span class="va">crc</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span>npcs <span class="op">=</span> <span class="fl">50</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span>,approx <span class="op">=</span> <span class="cn">F</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span>reduction <span class="op">=</span> <span class="st">"pca"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>,verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span>reduction <span class="op">=</span> <span class="st">"pca"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>,verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span>resolution <span class="op">=</span> <span class="fl">1.5</span>,verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/identity.html" class="external-link">identity</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric</span></span>
<span><span class="co">#&gt; To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'</span></span>
<span><span class="co">#&gt; This message will be shown once per session</span></span></code></pre></div>
<p>Here we directly use the cell type annotation we get from marker genes to annotate cell clusters.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cell_type</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">"../tests/data/cell_type.txt"</span>,header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">cell_type</span><span class="op">)</span> <span class="op">=</span> <span class="va">cell_type</span><span class="op">$</span><span class="va">cell</span></span>
<span><span class="va">crc</span><span class="op">$</span><span class="va">type</span> <span class="op">=</span> <span class="va">cell_type</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">crc</span><span class="op">)</span>,<span class="st">"type"</span><span class="op">]</span></span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">crc</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by <span class="op">=</span> <span class="st">"type"</span>,pt.size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="getStart_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="single-cell-heterogeneity-of-alterative-splicing">Single cell heterogeneity of alterative splicing<a class="anchor" aria-label="anchor" href="#single-cell-heterogeneity-of-alterative-splicing"></a>
</h3>
<p>Alternative splicing heterogeneity can only be confident estimated for genes with high expression in multiple cells, a common threshold is at least 10 transcripts in over 30 cells, thus we could first filter out low expressed genes via their total expression.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">heg</span> <span class="op">=</span> <span class="fu"><a href="../reference/HighExprsGene.html">HighExprsGene</a></span><span class="op">(</span><span class="va">crc</span>,thresh <span class="op">=</span> <span class="fl">500</span>,slot <span class="op">=</span> <span class="st">"RNA"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">heg</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "ACTB"    "ACTG1"   "AGR2"    "ALDOA"   "ANAPC11" "ANXA1"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">heg</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 355</span></span></code></pre></div>
<p>Here 355 genes can be preserved for downstream <span class="math inline">\(\phi\)</span> calculation. The gene bed and exon gtf annotations are optional but recommened. Gene bed annotations can be used to filter out signals from too short exons, which are volunarable to sequencing and mapping errors. Gtf annotation can help merge exons which are always coexisting or mutually exclusive, removing replicative signals.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">crc_phi</span> <span class="op">=</span> <span class="fu"><a href="../reference/genes_exons_phi.html">genes_exons_phi</a></span><span class="op">(</span><span class="va">crc</span>,genes <span class="op">=</span> <span class="va">heg</span>,gene_bed <span class="op">=</span> <span class="va">gene_bed</span>,gtf <span class="op">=</span> <span class="va">gtf</span>,cores <span class="op">=</span> <span class="fl">8</span><span class="op">)</span></span>
<span><span class="co">#&gt; The job will be paralleled on  8  cores.</span></span></code></pre></div>
<p>We can then show all valid <span class="math inline">\(\phi\)</span> estimation with small confidence interval in a <span class="math inline">\(\phi\)</span> vs. <span class="math inline">\(\bar{\psi}\)</span> scatter plot</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/phiPlot.html">phiPlot</a></span><span class="op">(</span><span class="va">crc_phi</span>,annot_col <span class="op">=</span> <span class="st">"gene"</span>,midpoint <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<p><img src="getStart_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
<p>We could order the <span class="math inline">\(\phi\)</span> for each exon in a decreasing order, and the top list should be highly variable exons.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">crc_phi</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">crc_phi</span><span class="op">$</span><span class="va">phi</span>,decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,<span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;       gene exon  mean_psi       phi   phi_lwr   phi_upr count  phi_conf</span></span>
<span><span class="co">#&gt; 367   MYL6   35 0.3684847 0.4236739 0.3453027 0.5096018    47 0.1642992</span></span>
<span><span class="co">#&gt; 1135  SAT1    9 0.2493341 0.3335311 0.2764232 0.3904058    78 0.1139826</span></span>
<span><span class="co">#&gt; 1136  SAT1   10 0.1751983 0.2399843 0.1944966 0.2958238    78 0.1013272</span></span>
<span><span class="co">#&gt; 399  NEAT1   39 0.5719830 0.2346518 0.1396089 0.3313117    36 0.1917028</span></span>
<span><span class="co">#&gt; 1141  SAT1   16 0.7338286 0.2153306 0.1561497 0.2639450    81 0.1077953</span></span>
<span><span class="co">#&gt; 1143  SAT1   18 0.7362526 0.2127588 0.1592439 0.2699973    81 0.1107534</span></span></code></pre></div>
<p>Here we use the highest one MYL6 exon35 as an example, we could first show its single cell <span class="math inline">\(\psi\)</span> distribution:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/psiHist.html">psiHist</a></span><span class="op">(</span><span class="va">crc</span>,gene <span class="op">=</span> <span class="st">"MYL6"</span>,exon <span class="op">=</span> <span class="st">"35"</span><span class="op">)</span></span></code></pre></div>
<p><img src="getStart_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/psiCellPlot.html">psiCellPlot</a></span><span class="op">(</span><span class="va">crc</span>,gene <span class="op">=</span> <span class="st">"MYL6"</span>,exons <span class="op">=</span> <span class="st">"35"</span>,cell_embedding <span class="op">=</span> <span class="st">"umap"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [[1]]</span></span></code></pre></div>
<p><img src="getStart_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
<p>We could see very different alternative splicing patterns for this exon in epithlials and other immune cells. Epithlials highly spliced in the exon 35 in MYL6, while T cells and macrophages tend to splice this exon out.</p>
</div>
<div class="section level3">
<h3 id="generalized-likelihood-ratio-test">Generalized Likelihood Ratio Test<a class="anchor" aria-label="anchor" href="#generalized-likelihood-ratio-test"></a>
</h3>
<p>Since we have found the splicing heterogeneity between epithelials and immune cells, here we could do a differential alternative splicig analysis between the two cell groups to see if they have other more splicing difference.</p>
<p>We first specify the non-epithelial cells as immune cells in the cell type annotation</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">crc</span><span class="op">$</span><span class="va">group</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">crc</span><span class="op">$</span><span class="va">type</span> <span class="op">==</span> <span class="st">"epithelial"</span>,<span class="st">"epithelial"</span>,<span class="st">"immune"</span><span class="op">)</span></span></code></pre></div>
<p>Then we could do generalized likelihood ratio test for highly expressed genes for immune cells and epithelials to see if they have alternative splicing difference.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">crc_GLRT_sig</span> <span class="op">=</span> <span class="fu"><a href="../reference/genes_groups_GLRT.html">genes_groups_GLRT</a></span><span class="op">(</span><span class="va">crc</span>,genes <span class="op">=</span> <span class="va">heg</span>,</span>
<span>                              group1s <span class="op">=</span> <span class="st">"epithelial"</span>,group2s <span class="op">=</span> <span class="st">"immune"</span>,group_col <span class="op">=</span> <span class="st">"group"</span>,</span>
<span>                              gene_bed <span class="op">=</span> <span class="va">gene_bed</span>,gtf <span class="op">=</span> <span class="va">gtf</span>,q_thresh <span class="op">=</span> <span class="fl">0.05</span>,cores <span class="op">=</span> <span class="fl">8</span><span class="op">)</span></span>
<span><span class="co">#&gt; The job will be paralleled on  8  cores.</span></span></code></pre></div>
<p>Compared to epthelials, most exons show decreasing splice-in in immune cells as most of them show negative mean change.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/GLRT_sig_plot.html">GLRT_sig_plot</a></span><span class="op">(</span><span class="va">crc_GLRT_sig</span><span class="op">)</span></span></code></pre></div>
<p><img src="getStart_files/figure-html/unnamed-chunk-20-1.png" width="700"></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">crc_GLRT_sig</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">crc_GLRT_sig</span><span class="op">$</span><span class="va">q</span><span class="op">)</span>,<span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;       gene     group1 group2 exon            p   alpha1    beta1   alpha2</span></span>
<span><span class="co">#&gt; 633   MYL6 epithelial immune   35 2.309757e-22 4.585580 3.364490 1.296922</span></span>
<span><span class="co">#&gt; 913  RPS24 epithelial immune   31 6.644542e-16 6.683290 3.766842 5.211762</span></span>
<span><span class="co">#&gt; 924   RPS8 epithelial immune   26 2.079360e-09 8.228405 4.128133 8.025231</span></span>
<span><span class="co">#&gt; 923   RPS8 epithelial immune   25 5.550707e-09 8.273600 4.109817 7.561798</span></span>
<span><span class="co">#&gt; 922   RPS8 epithelial immune   20 9.443566e-09 8.125215 4.073930 7.573987</span></span>
<span><span class="co">#&gt; 273 EIF4A1 epithelial immune   89 2.629369e-07 3.281194 4.891683 1.392122</span></span>
<span><span class="co">#&gt;        beta2            q  mean_diff      var_diff</span></span>
<span><span class="co">#&gt; 633 6.654382 2.556901e-19 -0.4136894 -0.0120241663</span></span>
<span><span class="co">#&gt; 913 7.757096 3.677754e-13 -0.2376737 -0.0029256764</span></span>
<span><span class="co">#&gt; 924 8.055760 7.672839e-07 -0.1668643 -0.0020203249</span></span>
<span><span class="co">#&gt; 923 7.435223 1.536158e-06 -0.1638993 -0.0009411618</span></span>
<span><span class="co">#&gt; 922 7.448730 2.090806e-06 -0.1618790 -0.0012499424</span></span>
<span><span class="co">#&gt; 273 6.878644 3.325090e-05 -0.2331552 -0.0110961237</span></span></code></pre></div>
<p>We could first check the <span class="math inline">\(\psi\)</span> distribution for MYL6 exon 35 in two cell groups as it’s the most significant signal.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/psiHist.html">psiHist</a></span><span class="op">(</span><span class="va">crc</span>,gene <span class="op">=</span> <span class="st">"MYL6"</span>,exon <span class="op">=</span> <span class="st">"35"</span>,group_col <span class="op">=</span> <span class="st">"group"</span>,group <span class="op">=</span> <span class="st">"epithelial"</span>,</span>
<span>           alpha <span class="op">=</span> <span class="fl">4.585</span>,beta <span class="op">=</span> <span class="fl">3.364</span><span class="op">)</span></span></code></pre></div>
<p><img src="getStart_files/figure-html/unnamed-chunk-22-1.png" width="700"></p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/psiHist.html">psiHist</a></span><span class="op">(</span><span class="va">crc</span>,gene <span class="op">=</span> <span class="st">"MYL6"</span>,exon <span class="op">=</span> <span class="st">"35"</span>,group_col <span class="op">=</span> <span class="st">"group"</span>,group <span class="op">=</span> <span class="st">"immune"</span>,</span>
<span>            alpha <span class="op">=</span> <span class="fl">1.296922</span>,beta <span class="op">=</span> <span class="fl">6.654382</span><span class="op">)</span></span></code></pre></div>
<p><img src="getStart_files/figure-html/unnamed-chunk-22-2.png" width="700"></p>
<p>We could also check other genes, for example, RPS24:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/psiHist.html">psiHist</a></span><span class="op">(</span><span class="va">crc</span>,gene <span class="op">=</span> <span class="st">"RPS24"</span>,exon <span class="op">=</span> <span class="st">"31"</span>,group_col <span class="op">=</span> <span class="st">"group"</span>,group <span class="op">=</span> <span class="st">"epithelial"</span>,</span>
<span>           alpha <span class="op">=</span> <span class="fl">6.683290</span>,beta <span class="op">=</span> <span class="fl">3.766842</span><span class="op">)</span></span></code></pre></div>
<p><img src="getStart_files/figure-html/unnamed-chunk-23-1.png" width="700"></p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/psiHist.html">psiHist</a></span><span class="op">(</span><span class="va">crc</span>,gene <span class="op">=</span> <span class="st">"RPS24"</span>,exon <span class="op">=</span> <span class="st">"31"</span>,group_col <span class="op">=</span> <span class="st">"group"</span>,group <span class="op">=</span> <span class="st">"immune"</span>,</span>
<span>            alpha <span class="op">=</span> <span class="fl">5.211762</span>,beta <span class="op">=</span> <span class="fl">7.757096</span><span class="op">)</span></span></code></pre></div>
<p><img src="getStart_files/figure-html/unnamed-chunk-23-2.png" width="700"></p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Yuntian Fu.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
