<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Longcell">
<title>Tutorial for Longcell â€¢ Longcell</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Tutorial for Longcell">
<meta property="og:description" content="Longcell">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">Longcell</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/Longcell.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Tutorial for Longcell</h1>
            
      
      
      <div class="d-none name"><code>Longcell.Rmd</code></div>
    </div>

    
    
<div class="section level3">
<h3 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h3>
<p>This tutorial demonstrates how to use Longcell to analyze single cell
isoform count data. Due to the frequent truncations in Nanopore long
reads sequencing and to allow for identification of new ioforms,
Longcell stores the isoform information as sequence of splicing sites.
So instead of building a cell-by-isoform matrix, Longcell reveal
alternative splicing events on the splicing sites level. But the
cell-by-isoform matrix is also provided for regular single cell RNA seq
analysis,like cell type clustering.</p>
<p>The general steps for the analysis are:</p>
<ol style="list-style-type: decimal">
<li><p>regular scRNA-seq gene/isoform expression analysis.</p></li>
<li><p>identify highly variable splicing sites.</p></li>
<li><p>differetial alternative splicing analysis between cell
groups</p></li>
</ol>
<p>First, we load Longcell and the other packages necessary for this
vignette.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>future.globals.maxSize<span class="op">=</span> <span class="fl">1024</span><span class="op">*</span><span class="fl">1024</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://yuntianf.github.io/Longcell/">Longcell</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: SeuratObject</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: sp</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'SeuratObject'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     intersect</span></span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'dplyr'</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:stats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     filter, lag</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     intersect, setdiff, setequal, union</span></span></code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://future.futureverse.org" class="external-link">future</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://future.apply.futureverse.org" class="external-link">future.apply</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="annotation">annotation<a class="anchor" aria-label="anchor" href="#annotation"></a>
</h3>
<p>The gene strand information is necessary to build meta splicing sites
in the downstream analysis, here we can use the
<code>gene_bed.rds</code> output from the LongcellPre package, but you
can also use your annotation file to indicate the gene strand
information. It should be a table with at least two columns, one for the
gene name, and the other one for the gene strand.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene_bed</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"../inst/extdata//annotation//gene_bed.rds"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="data-overview">data overview<a class="anchor" aria-label="anchor" href="#data-overview"></a>
</h3>
<p>Here we are using the sample of colorecatal metastasis to liver
published in our paper as an example. The single cell isoform
quantification for this sample is generated by LongcellPre.</p>
<p>We can generate a Seurat object based on the cell by isoform matrix,
just in the same way as the cell by gene matrix.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">iso_mat</span> <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/Read10X.html" class="external-link">Read10X</a></span><span class="op">(</span><span class="st">"../inst/extdata/data/isoform/"</span>,gene.column <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">crc</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">iso_mat</span>,project <span class="op">=</span> <span class="st">"isoform"</span>,min.cells <span class="op">=</span> <span class="fl">3</span>,min.features <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Feature names cannot have underscores ('_'), replacing with dashes</span></span>
<span><span class="co">## ('-')</span></span></code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">crc</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 17767 features across 236 samples within 1 assay </span></span>
<span><span class="co">## Active assay: RNA (17767 features, 0 variable features)</span></span>
<span><span class="co">##  1 layer present: counts</span></span></code></pre>
<p>We can first do general single cell analysis via regular Seurat
process to get the cell cluster information</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="op">{</span></span>
<span>    <span class="va">crc</span> <span class="op">&lt;-</span>  <span class="va">crc</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://satijalab.org/seurat/reference/SCTransform.html" class="external-link">SCTransform</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"glmGamPoi"</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span>npcs <span class="op">=</span> <span class="fl">30</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span>reduction <span class="op">=</span> <span class="st">"pca"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span>,n.neighbors <span class="op">=</span> <span class="fl">30</span>,verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Found more than one class "dist" in cache; using the first, from namespace 'spam'</span></span></code></pre>
<pre><code><span><span class="co">## Also defined by 'BiocGenerics'</span></span></code></pre>
<pre><code><span><span class="co">## Found more than one class "dist" in cache; using the first, from namespace 'spam'</span></span></code></pre>
<pre><code><span><span class="co">## Also defined by 'BiocGenerics'</span></span></code></pre>
<p>Here we directly use the cell type annotation we get from marker
genes to annotate cell clusters.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cell_type</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">"../inst/extdata/data/cell_type.txt"</span>,header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">cell_type</span><span class="op">)</span> <span class="op">=</span> <span class="va">cell_type</span><span class="op">$</span><span class="va">cell</span></span>
<span><span class="va">crc</span><span class="op">$</span><span class="va">type</span> <span class="op">=</span> <span class="va">cell_type</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">crc</span><span class="op">)</span>,<span class="st">"type"</span><span class="op">]</span></span></code></pre></div>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">crc</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by <span class="op">=</span> <span class="st">"type"</span>,label <span class="op">=</span> <span class="cn">TRUE</span>,pt.size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="Longcell_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="single-cell-heterogeneity-of-alterative-splicing">Single cell heterogeneity of alterative splicing<a class="anchor" aria-label="anchor" href="#single-cell-heterogeneity-of-alterative-splicing"></a>
</h3>
<p>After we take a glance on the data, we can start to explore the
alternative splicing within the cell population. Longcell detect the
alternative splicing event at the splice site level.We first read in the
data for the downstream analysis.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">iso</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"../inst/extdata/data/sc_iso_count.rds"</span><span class="op">)</span></span></code></pre></div>
<p>Each read here is stored as <span class="math inline">\(s_1,e_1|s_2,e_2|...\)</span>, in which <span class="math inline">\(s_i\)</span> means the start position of the exon
<span class="math inline">\(i\)</span> while <span class="math inline">\(e_i\)</span> means the end position.</p>
<p>We first build a Splice object based on this data.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">crc_splice</span> <span class="op">=</span> <span class="fu"><a href="../reference/creatSplice_from_df.html">creatSplice_from_df</a></span><span class="op">(</span><span class="va">iso</span><span class="op">)</span></span></code></pre></div>
<p>The isoform sequences and expression can be revisted from the Splice
object by <code><a href="../reference/getIsoform.html">getIsoform()</a></code></p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">temp</span> <span class="op">=</span> <span class="fu"><a href="../reference/getIsoform.html">getIsoform</a></span><span class="op">(</span><span class="va">crc_splice</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MYL6"</span>,<span class="st">"RPS24"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">temp</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                    cell  gene</span></span>
<span><span class="co">## 575755 AAACGGGGTTATGTGC RPS24</span></span>
<span><span class="co">## 575756 AAACGGGGTTATGTGC RPS24</span></span>
<span><span class="co">## 575757 AAACGGGGTTATGTGC RPS24</span></span>
<span><span class="co">## 575758 AAACGGGGTTATGTGC RPS24</span></span>
<span><span class="co">## 575759 AAACGGGGTTATGTGC RPS24</span></span>
<span><span class="co">## 575760 AAACGGGGTTATGTGC RPS24</span></span>
<span><span class="co">##                                                                                                            isoform</span></span>
<span><span class="co">## 575755                   78033859,78033904|78035352,78035417|78035511,78035720|78037194,78037304|78040615,78040692</span></span>
<span><span class="co">## 575756 78033863,78033904|78035352,78035417|78035511,78035720|78037194,78037304|78040204,78040225|78040615,78035580</span></span>
<span><span class="co">## 575757 78033863,78033904|78035352,78035417|78035511,78035720|78037194,78037304|78040204,78040225|78040615,78040694</span></span>
<span><span class="co">## 575758 78033863,78033904|78035352,78035417|78035511,78035720|78037194,78037304|78040204,78040225|78040615,78040695</span></span>
<span><span class="co">## 575759 78033863,78033904|78035352,78035417|78035511,78035720|78037194,78037304|78040204,78040225|78040615,78040696</span></span>
<span><span class="co">## 575760 78033863,78033904|78035352,78035417|78035511,78035720|78037194,78037304|78040204,78040225|78040615,78040697</span></span>
<span><span class="co">##            count     polyA</span></span>
<span><span class="co">## 575755 0.1904762 1.0000000</span></span>
<span><span class="co">## 575756 0.3333333 0.0000000</span></span>
<span><span class="co">## 575757 3.3333333 0.9914286</span></span>
<span><span class="co">## 575758 1.3333333 1.0000000</span></span>
<span><span class="co">## 575759 1.6666667 0.9898305</span></span>
<span><span class="co">## 575760 6.0000000 0.9949483</span></span></code></pre>
<p>As isoform count is even sparser than the gene count, we can only get
confident alternative splicing signals from high expression genes, thus
we can filter out lowly expressed genes to save the computing time.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">heg</span> <span class="op">=</span> <span class="fu"><a href="../reference/HighExprsGene.html">HighExprsGene</a></span><span class="op">(</span><span class="va">crc_splice</span>,thresh <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">heg</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "ABRACL" "ACTB"   "ACTG1"  "ACTN4"  "AGR2"   "AIF1"</span></span></code></pre>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">heg</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 752</span></span></code></pre>
<p>Here 752 genes would be preserved for downstream <span class="math inline">\(\phi\)</span> calculation. And before that, we can
first merge splicing sites which are always coexisting or mutually
exclusive as meta splicing sites, as they contain the same information.
Through this step we can further decrease the computation burden. The
parameter <code>eps</code> can be adjusted to indicate how aggressive
the merging process would be.Higher <code>eps</code> leads to more
aggressive merge, but may be more possible to lead to information
loss.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span>strategy <span class="op">=</span> <span class="st">"multisession"</span>,workers <span class="op">=</span> <span class="fl">8</span><span class="op">)</span></span>
<span><span class="va">crc_splice</span> <span class="op">=</span> <span class="fu"><a href="../reference/geneSiteTable.html">geneSiteTable</a></span><span class="op">(</span><span class="va">crc_splice</span>,<span class="va">gene_bed</span>,genes <span class="op">=</span> <span class="va">heg</span>,eps <span class="op">=</span> <span class="fl">0.05</span>,verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>The merged splicing sites are saved in a <code>metaSite</code>
object, for example:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/getMetaSites.html">getMetaSites</a></span><span class="op">(</span><span class="va">crc_splice</span>,<span class="st">"MYL6"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## metaSite object with  225  cells and  4 meta splicing sites, which are merged from  25 original splicing sites.</span></span>
<span><span class="co">## The head of the gene count matrix for each meta site is:</span></span>
<span><span class="co">##                  metasite_1 metasite_2 metasite_3 metasite_4</span></span>
<span><span class="co">## AAACGGGGTTATGTGC         29         28         29         29</span></span>
<span><span class="co">## AACCATGAGCCACGCT          5          4          5          5</span></span>
<span><span class="co">## AACCATGCAAGCCTAT          6          5          5          5</span></span>
<span><span class="co">## AACTCTTCAAGTAATG          3          2          3          2</span></span>
<span><span class="co">## AAGACCTAGCTGAAAT          2          1          2          2</span></span>
<span><span class="co">## AAGACCTTCAAACGGG         12          9         12         12</span></span>
<span><span class="co">## The head of the spliced in count matrix for each meta site is:</span></span>
<span><span class="co">##                  metasite_1 metasite_2 metasite_3 metasite_4</span></span>
<span><span class="co">## AAACGGGGTTATGTGC         29         28         29          2</span></span>
<span><span class="co">## AACCATGAGCCACGCT          5          4          5          0</span></span>
<span><span class="co">## AACCATGCAAGCCTAT          6          5          5          0</span></span>
<span><span class="co">## AACTCTTCAAGTAATG          3          2          2          2</span></span>
<span><span class="co">## AAGACCTAGCTGAAAT          2          1          2          1</span></span>
<span><span class="co">## AAGACCTTCAAACGGG         12          9         12          7</span></span></code></pre>
<p>The <code>metaSite</code> object maintains two matrices, one records
the gene count for each meta splice site and the other records the
spliced-in count. The correspondence between the meta sites and original
sites is also recorded in it.</p>
<p>After we generate the meta splicing sites, we can estimate the ratio
of inter-cell heterogeneity for each meta site.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span>strategy <span class="op">=</span> <span class="st">"multisession"</span>,workers <span class="op">=</span> <span class="fl">8</span><span class="op">)</span></span>
<span><span class="va">crc_phi</span> <span class="op">=</span> <span class="fu"><a href="../reference/genesSitesPhi.html">genesSitesPhi</a></span><span class="op">(</span><span class="va">crc_splice</span>,genes <span class="op">=</span> <span class="va">heg</span>,method <span class="op">=</span> <span class="st">"beta"</span>,verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>Each row in the result table records the <span class="math inline">\(\bar{\psi}\)</span> (percent-spliced-in) and <span class="math inline">\(\phi\)</span> (inter-cell heterogeneity)
estimation and the confidence interval for <span class="math inline">\(\phi\)</span>. The count <code>column</code> means
the number of cells which have valid gene expression to estimate the
<span class="math inline">\(\psi\)</span> and <span class="math inline">\(\phi\)</span>.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">crc_phi</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      gene meta_sites  mean_psi        phi      phi_lwr   phi_upr count</span></span>
<span><span class="co">## 1    AGR2 metasite_2 0.7553166 0.05558393 -0.007364070 0.1185319    20</span></span>
<span><span class="co">## 2 ANAPC11 metasite_2 0.4206358 0.15273095  0.060189270 0.2452726    28</span></span>
<span><span class="co">## 3 ANAPC11 metasite_6 0.4774544 0.14312845  0.050723033 0.2355339    28</span></span>
<span><span class="co">## 4    AREG metasite_2 0.8635209 0.10122846  0.002846471 0.1996104    23</span></span>
<span><span class="co">## 5    ATF3 metasite_6 0.7075183 0.13312699  0.040007757 0.2262462    21</span></span>
<span><span class="co">## 6  BCL2A1 metasite_2 0.1698188 0.09520912  0.021722143 0.1686961    32</span></span>
<span><span class="co">##    phi_conf</span></span>
<span><span class="co">## 1 0.1258960</span></span>
<span><span class="co">## 2 0.1850834</span></span>
<span><span class="co">## 3 0.1848108</span></span>
<span><span class="co">## 4 0.1967640</span></span>
<span><span class="co">## 5 0.1862385</span></span>
<span><span class="co">## 6 0.1469739</span></span></code></pre>
<p>We can then show all valid <span class="math inline">\(\phi\)</span>
estimation with small confidence interval in a <span class="math inline">\(\phi\)</span> vs.Â <span class="math inline">\(\bar{\psi}\)</span> scatter plot</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/phiPlot.html">phiPlot</a></span><span class="op">(</span><span class="va">crc_phi</span>,annot_col <span class="op">=</span> <span class="st">"gene"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Longcell_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
<p>We could order the <span class="math inline">\(\phi\)</span> for each
meta splicing site in a decreasing order, and the top list should be
highly variable exons.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">crc_phi</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">crc_phi</span><span class="op">$</span><span class="va">phi</span>,decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,<span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##     gene meta_sites  mean_psi       phi    phi_lwr   phi_upr count   phi_conf</span></span>
<span><span class="co">## 41  MYL6 metasite_4 0.3110294 0.2979716 0.23732462 0.3586186   111 0.12129397</span></span>
<span><span class="co">## 66  SAT1 metasite_3 0.6718207 0.2067908 0.16702254 0.2465591   157 0.07953661</span></span>
<span><span class="co">## 25  IER3 metasite_2 0.8422321 0.1924780 0.10616716 0.2787887    53 0.17262158</span></span>
<span><span class="co">## 63 RPS24 metasite_3 0.6695511 0.1822913 0.15178788 0.2127947   221 0.06100680</span></span>
<span><span class="co">## 74 SNHG6 metasite_2 0.8348119 0.1676768 0.08307809 0.2522754    44 0.16919734</span></span>
<span><span class="co">## 16  ELOC metasite_4 0.6006515 0.1593375 0.06683127 0.2518437    34 0.18501247</span></span></code></pre>
<p>Here we use the highest one MYL6 metasite_4 as an example, we could
first show its single cell <span class="math inline">\(\psi\)</span>
distribution,the color of the bar indicates the average gene
expression:</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene</span> <span class="op">=</span> <span class="st">"MYL6"</span></span>
<span><span class="va">site</span> <span class="op">=</span> <span class="st">"metasite_4"</span></span></code></pre></div>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/psiHist.html">psiHist</a></span><span class="op">(</span><span class="va">crc_splice</span>,gene <span class="op">=</span> <span class="va">gene</span>,site <span class="op">=</span> <span class="va">site</span><span class="op">)</span></span></code></pre></div>
<p><img src="Longcell_files/figure-html/unnamed-chunk-22-1.png" width="700"></p>
<p>We could also show the <span class="math inline">\(\psi\)</span>
distribution in the cell embedding umap:</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">embedding</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">crc</span><span class="op">@</span><span class="va">reductions</span><span class="op">$</span><span class="va">umap</span><span class="op">@</span><span class="va">cell.embeddings</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/psiCellPlot.html">psiCellPlot</a></span><span class="op">(</span><span class="va">crc_splice</span>,gene <span class="op">=</span> <span class="va">gene</span>,site <span class="op">=</span> <span class="va">site</span>,cell_embedding <span class="op">=</span> <span class="va">embedding</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [[1]]</span></span></code></pre>
<p><img src="Longcell_files/figure-html/unnamed-chunk-24-1.png" width="700"></p>
<p>We could see very different alternative splicing patterns for this
exon in epithlials and other immune cells. Epithlials highly spliced in
the metasite_4 in MYL6, while T cells and macrophages tend to splice
this meta site out.</p>
<p>If we trace back to the original sites, we could find that the
metasite_4 contains two sites:</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/getMetaSites.html">getMetaSites</a></span><span class="op">(</span><span class="va">crc_splice</span>,<span class="st">"MYL6"</span><span class="op">)</span><span class="op">@</span><span class="va">sites</span></span></code></pre></div>
<pre><code><span><span class="co">## [[1]]</span></span>
<span><span class="co">##  [1] "56158404_out" "56158415_out" "56158684_in"  "56158711_out" "56158902_in" </span></span>
<span><span class="co">##  [6] "56158920_out" "56159498_in"  "56159587_in"  "56159622_out" "56159975_in" </span></span>
<span><span class="co">## [11] "56160025_in"  "56160055_out" "56160148_out" "56160243_in"  "56160284_out"</span></span>
<span><span class="co">## [16] "56161465_out" "56161547_in"  "56162264_in"  "56162331_out" "56162855_in" </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[2]]</span></span>
<span><span class="co">## [1] "56159730_out"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[3]]</span></span>
<span><span class="co">## [1] "56160320_out" "56161387_in" </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[4]]</span></span>
<span><span class="co">## [1] "56160626_in"  "56160670_out"</span></span></code></pre>
<p>We can compare the sashimi plot between the epithlials and the immune
cells group to see which exons are differetially spliced between the two
cell groups.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">epithlials</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">crc</span><span class="op">)</span><span class="op">[</span><span class="va">crc</span><span class="op">$</span><span class="va">type</span> <span class="op">==</span> <span class="st">"epithelial"</span><span class="op">]</span></span>
<span><span class="va">immune</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">crc</span><span class="op">)</span><span class="op">[</span><span class="va">crc</span><span class="op">$</span><span class="va">type</span> <span class="op">!=</span> <span class="st">"epithelial"</span><span class="op">]</span></span></code></pre></div>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/sashimi.html">sashimi</a></span><span class="op">(</span><span class="va">crc_splice</span>,gene <span class="op">=</span> <span class="st">"MYL6"</span>,cells <span class="op">=</span> <span class="va">epithlials</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">xlim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">56158300</span>,<span class="fl">56161500</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Removed 1009 rows containing missing values (`geom_segment()`).</span></span></code></pre>
<p><img src="Longcell_files/figure-html/unnamed-chunk-27-1.png" width="700"></p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/sashimi.html">sashimi</a></span><span class="op">(</span><span class="va">crc_splice</span>,gene <span class="op">=</span> <span class="st">"MYL6"</span>,cells <span class="op">=</span> <span class="va">immune</span>,color_id <span class="op">=</span> <span class="fl">2</span>,filter_ratio <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">+</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">xlim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">56158300</span>,<span class="fl">56161500</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Removed 1719 rows containing missing values (`geom_segment()`).</span></span></code></pre>
<p><img src="Longcell_files/figure-html/unnamed-chunk-28-1.png" width="700"></p>
<p>From the sashimi plot we can see that the exon <span class="math inline">\(6\)</span> has differetial alternative splicing
between the epithlials and immune cells.</p>
</div>
<div class="section level3">
<h3 id="generalized-likelihood-ratio-test">Generalized Likelihood Ratio Test<a class="anchor" aria-label="anchor" href="#generalized-likelihood-ratio-test"></a>
</h3>
<p>Since we have found the splicing heterogeneity between epithelials
and immune cells, here we could do a differential alternative splicig
analysis between the two cell groups to see if they have other more
splicing difference.</p>
<p>Still, we would do generalized likelihood ratio test for highly
expressed genes.</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span>strategy <span class="op">=</span> <span class="st">"multisession"</span>,workers <span class="op">=</span> <span class="fl">8</span><span class="op">)</span></span>
<span><span class="va">group1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">epithlials</span><span class="op">)</span></span>
<span><span class="va">group2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">immune</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">group1</span><span class="op">)</span> <span class="op">=</span> <span class="st">"epithlials"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">group2</span><span class="op">)</span> <span class="op">=</span> <span class="st">"immune"</span></span>
<span><span class="va">crc_GLRT_sig</span> <span class="op">=</span> <span class="fu"><a href="../reference/genes_groups_GLRT.html">genes_groups_GLRT</a></span><span class="op">(</span><span class="va">crc_splice</span>,genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">crc_splice</span><span class="op">@</span><span class="va">meta_sites</span><span class="op">)</span>,</span>
<span>                              group1s <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">epithlials</span><span class="op">)</span>,group2s <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">immune</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "optimization failed!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span></code></pre>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span>strategy <span class="op">=</span> <span class="st">"multisession"</span>,workers <span class="op">=</span> <span class="fl">8</span><span class="op">)</span></span>
<span><span class="va">crc_GLRT_sig</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="fu"><a href="../reference/genes_multigroups_GLRT.html">genes_multigroups_GLRT</a></span><span class="op">(</span>spliceOb <span class="op">=</span> <span class="va">crc_splice</span>,</span>
<span>                                      genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">crc_splice</span><span class="op">@</span><span class="va">meta_sites</span><span class="op">)</span>,</span>
<span>                                      meta <span class="op">=</span> <span class="va">cell_type</span>,group_col <span class="op">=</span> <span class="st">"type"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "optimization failed!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "The MLE based on current starting points doesn't converge, will try other starting points!"</span></span>
<span><span class="co">## [1] "optimization failed!"</span></span></code></pre>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">crc_GLRT_sig</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">crc_GLRT_sig</span><span class="op">$</span><span class="va">q</span><span class="op">)</span>,<span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##           group  gene       site            p     alpha1       beta1     alpha2</span></span>
<span><span class="co">## 781  epithelial RPS24 metasite_3 4.594592e-28 15.4360341   2.3924591  4.4251506</span></span>
<span><span class="co">## 836  epithelial  MYL6 metasite_4 1.034219e-26  2.6446410   2.2051834  1.8689681</span></span>
<span><span class="co">## 823  epithelial PCBP2 metasite_3 1.155557e-10  0.8140863   0.5537612 20.2299469</span></span>
<span><span class="co">## 690  epithelial SSBP1 metasite_1 3.452990e-10 12.6321858   4.5543177  0.5977454</span></span>
<span><span class="co">## 300 macrophages SERF2 metasite_2 4.188171e-09  3.8181003 104.3291222  1.6117449</span></span>
<span><span class="co">## 920  epithelial SERF2 metasite_2 1.069562e-08  2.5073500  15.8760372  1.1346810</span></span>
<span><span class="co">##          beta2        emd            q    mean_diff     var_diff</span></span>
<span><span class="co">## 781  3.2491754 0.28918950 6.960806e-25 -0.289189500  0.021973232</span></span>
<span><span class="co">## 836 11.1277555 0.40150353 7.834208e-24 -0.401503529 -0.033588820</span></span>
<span><span class="co">## 823 32.2189723 0.28745582 5.835564e-08 -0.209451054 -0.097323939</span></span>
<span><span class="co">## 690  0.2075395 0.19931300 1.307820e-07  0.007272022  0.095257662</span></span>
<span><span class="co">## 300 13.2723842 0.07304176 1.269016e-06  0.072981485  0.005767000</span></span>
<span><span class="co">## 920 18.3328958 0.07810646 2.700643e-06 -0.078106482 -0.003395092</span></span></code></pre>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/GLRT_sig_plot.html">GLRT_sig_plot</a></span><span class="op">(</span><span class="va">crc_GLRT_sig</span>,color_col <span class="op">=</span> <span class="st">"group"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Longcell_files/figure-html/unnamed-chunk-32-1.png" width="700"></p>
<p>The metasite_4 in MYL6 is one of the most significant result in the
differential alternative splicing analysis between the immune and
epithelial cells, which follows our expectation. As from last section,
we do observe obvious inter-cell heterogeneity for this metasite.</p>
<p>Besides MYL6, there also other metasites show siginificant
difference. For example, the metasite_3 in RPS24, we can observe its
<span class="math inline">\(\psi\)</span> distribtion in two cell
populations as a sanity check.</p>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">target_group</span> <span class="op">=</span> <span class="st">"epithelial"</span></span>
<span><span class="va">target_gene</span> <span class="op">=</span> <span class="st">"RPS24"</span></span>
<span><span class="va">target_site</span> <span class="op">=</span> <span class="st">"metasite_3"</span></span>
<span><span class="va">target</span> <span class="op">=</span> <span class="va">crc_GLRT_sig</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">group</span> <span class="op">==</span> <span class="va">target_group</span>,<span class="va">gene</span> <span class="op">==</span> <span class="va">target_gene</span>,<span class="va">site</span> <span class="op">==</span> <span class="va">target_site</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">target_cells</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">cell_type</span><span class="op">)</span><span class="op">[</span><span class="va">cell_type</span><span class="op">$</span><span class="va">type</span> <span class="op">==</span> <span class="va">target_group</span><span class="op">]</span></span>
<span><span class="va">control_cells</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">cell_type</span><span class="op">)</span><span class="op">[</span><span class="va">cell_type</span><span class="op">$</span><span class="va">type</span> <span class="op">!=</span> <span class="va">target_group</span><span class="op">]</span></span></code></pre></div>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/psiHist.html">psiHist</a></span><span class="op">(</span><span class="va">crc_splice</span>,gene <span class="op">=</span> <span class="va">target_gene</span>,site <span class="op">=</span> <span class="va">target_site</span>,cells <span class="op">=</span> <span class="va">target_cells</span>,</span>
<span>        alpha <span class="op">=</span> <span class="va">target</span><span class="op">$</span><span class="va">alpha1</span>,beta <span class="op">=</span> <span class="va">target</span><span class="op">$</span><span class="va">beta1</span><span class="op">)</span></span></code></pre></div>
<p><img src="Longcell_files/figure-html/unnamed-chunk-35-1.png" width="700"></p>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/psiHist.html">psiHist</a></span><span class="op">(</span><span class="va">crc_splice</span>,gene <span class="op">=</span> <span class="va">target_gene</span>,site <span class="op">=</span> <span class="va">target_site</span>,cells <span class="op">=</span> <span class="va">control_cells</span>,</span>
<span>        alpha <span class="op">=</span> <span class="va">target</span><span class="op">$</span><span class="va">alpha2</span>,beta <span class="op">=</span> <span class="va">target</span><span class="op">$</span><span class="va">beta2</span><span class="op">)</span></span></code></pre></div>
<p><img src="Longcell_files/figure-html/unnamed-chunk-35-2.png" width="700"></p>
<p>From above histogram we do see, compared to the immune cells, RPS24
in epthelial preserved more metasite_3.</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/getMetaSites.html">getMetaSites</a></span><span class="op">(</span><span class="va">crc_splice</span>,<span class="va">target_gene</span><span class="op">)</span><span class="op">@</span><span class="va">sites</span></span></code></pre></div>
<pre><code><span><span class="co">## [[1]]</span></span>
<span><span class="co">##  [1] "78033904_out" "78033986_out" "78034232_in"  "78034316_out" "78034421_out"</span></span>
<span><span class="co">##  [6] "78035352_in"  "78035417_out" "78035511_in"  "78035646_in"  "78037219_in" </span></span>
<span><span class="co">## [11] "78037290_out" "78037965_in"  "78037982_out" "78040629_in"  "78040647_in" </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[2]]</span></span>
<span><span class="co">## [1] "78035720_out" "78037194_in"  "78037304_out" "78040615_in" </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[3]]</span></span>
<span><span class="co">## [1] "78040204_in"  "78040225_out"</span></span></code></pre>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/sashimi.html">sashimi</a></span><span class="op">(</span><span class="va">crc_splice</span>,gene <span class="op">=</span> <span class="va">target_gene</span>,cells <span class="op">=</span> <span class="va">target_cells</span><span class="op">)</span></span></code></pre></div>
<p><img src="Longcell_files/figure-html/unnamed-chunk-37-1.png" width="700"></p>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/sashimi.html">sashimi</a></span><span class="op">(</span><span class="va">crc_splice</span>,gene <span class="op">=</span> <span class="va">target_gene</span>,cells <span class="op">=</span> <span class="va">control_cells</span>,color_id <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="Longcell_files/figure-html/unnamed-chunk-38-1.png" width="700"></p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Yuntian Fu.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
